orchagent/flex_counter/flowcounterrouteorch.cpp

```diff
@@ -0,0 +1,985 @@
+    if (insert_result.second)
+    {
+        SWSS_LOG_NOTICE("Inserting route pattern %s, max match count is %zu", pattern.c_str(), max_match_count);
+        if (!validateRoutePattern(*insert_result.first))
+        {
+            mRoutePatternSet.erase(insert_result.first);
+            return;
+        }
+
+        createRouteFlowCounterByPattern(*insert_result.first, 0);
+    }
+    else
+    {
+        SWSS_LOG_NOTICE("Updating route pattern %s max match count to %zu", pattern.c_str(), max_match_count);
+        RoutePattern &existing = const_cast<RoutePattern &>(*insert_result.first);
+        onRoutePatternMaxMatchCountChange(existing, max_match_count);
+    }
+}
+
+void FlowCounterRouteOrch::removeRoutePattern(const std::string& pattern)
+{
+    SWSS_LOG_ENTER();
+    sai_object_id_t vrf_id;
+    IpPrefix ip_prefix;
+    std::string vrf_name;
+    if (!parseRouteKeyForRoutePattern(pattern, '|', vrf_id, ip_prefix, vrf_name))
+    {
+        vrf_id = SAI_NULL_OBJECT_ID;
+    }
+
+    SWSS_LOG_NOTICE("Removing route pattern %s", pattern.c_str());
+    RoutePattern route_pattern(vrf_name, vrf_id, ip_prefix, 0);
+    auto iter = mRoutePatternSet.find(route_pattern);
+    if (iter == mRoutePatternSet.end())
+    {
+        // Should not go to this branch, just in case
+        SWSS_LOG_ERROR("Trying to remove route pattern %s, but it does not exist", pattern.c_str());
+        return;
+    }
+    mRoutePatternSet.erase(iter);
+
+    removeRoutePattern(route_pattern);
+}
+
+void FlowCounterRouteOrch::removeRoutePattern(const RoutePattern &route_pattern)
+{
+    SWSS_LOG_ENTER();
+    auto cache_iter = mBoundRouteCounters.find(route_pattern);
+    if (cache_iter != mBoundRouteCounters.end())
+    {
+        for (auto &entry : cache_iter->second)
+        {
+            removeRouteFlowCounterFromDB(route_pattern.vrf_id, entry.first, entry.second);
+            unbindFlowCounter(route_pattern, route_pattern.vrf_id, entry.first, entry.second);
+        }
+        mBoundRouteCounters.erase(cache_iter);
+    }
+
+    auto pending_iter = mPendingAddToFlexCntr.find(route_pattern);
+    if (pending_iter != mPendingAddToFlexCntr.end())
+    {
+        for (auto &entry : pending_iter->second)
+        {
+            unbindFlowCounter(route_pattern, route_pattern.vrf_id, entry.first, entry.second);
+        }
+        mPendingAddToFlexCntr.erase(pending_iter);
+    }
+}
+
+void FlowCounterRouteOrch::onAddMiscRouteEntry(sai_object_id_t vrf_id, const sai_ip_prefix_t& ip_pfx, bool add_to_cache)
+{
+    SWSS_LOG_ENTER();
+    if (!mRouteFlowCounterSupported)
+    {
+        return;
+    }
+
+    IpPrefix ip_prefix = convert(ip_pfx);
+    onAddMiscRouteEntry(vrf_id, ip_prefix, add_to_cache);
+}
+
+void FlowCounterRouteOrch::onAddMiscRouteEntry(sai_object_id_t vrf_id, const IpPrefix& ip_prefix, bool add_to_cache)
+{
+    SWSS_LOG_ENTER();
+    if (!mRouteFlowCounterSupported)
```

Comment:
I think there is a gap here. My point was this code flow is executed only if `m_route_flow_counter_enabled` is true. Not just platform support.  The idea was, there is no change to the existing flow, if the feature is disabled.  We should check for both cases here.  `!mRouteFlowCounterSupported || !m_route_flow_counter_enabled`

Reply:
We have to save it here. For example:
1. Some misc route entries are created, and we don't save it here. In addition, those misc route entries are not in DB/memory.
2. User enables route flow counter and configures route pattern
3. flexcounterrouteorch starts to search marched route entries, but misc route entries were not saved anywhere, hence it can never be matched.

Reply:
Only case of misc route is IP2ME and Vnet routes. For this, I don't think we should cache the entire entries if the feature is never enabled. As mentioned, this is a debug feature.

Reply:
Hi Prince, we don't cache VNET routes in flowcounterrouteorch, we call it like this:

```cpp
gFlowCounterRouteOrch->onRemoveMiscRouteEntry(vr_id, ip_pfx, false);
```

The last argument "false" indicates that we don't cache the routes.

But we do cache IP2ME routes. Like I said before, we have to do it because IP2ME routes are not saved to any memory data structures. So route pattern can never match an IP2ME routes if those routes are created before enabling route flow counter (no place to search from). IMHO, we need to cache IP2ME routes or we have to claim that we don't support IP2ME route for this counter, however, it was defined in the HLD.